### PR TITLE
feat: Add Array type to MultipartFormDataType variants for REST api

### LINKED ISSUE #10876: [Feature]: Allow dynamic string arrays in multipart
### Is there an existing issue for this?

- [x] I have searched the existing issues

### Summary

APIs using multipart would commonly expect list of values to be sent as unique values using the same key;

```
// My source data could be an output of a Select for example with value ['a', 'b', 'c']

// In JSON we would have something like this
{
  myList: ['a', 'b', 'c']
}

// In multipart the request might look like this
--12345
Content-Disposition: form-data; name="myList"

a
--12345
Content-Disposition: form-data; name="myList"

b
--12345
Content-Disposition: form-data; name="myList"

c
--12345
```

Currently it is not possible to map an array value in such a way in appsmith

### Why should this be worked on?

APIs expecting requests using this format can not be achieved, unless there is a fixed and contact length array to send.

### PR DESCRIPTION
**Pull request by mojtab23**
## Description
Users can add an array as a value for multipart requests, and it will handle every item in the list as a part of the multipart request.

Screenshot 2022-03-14 110525

Example body:
```
--IpJ9ACFxVbhXlAQuIFccoJ0K0ttFS5PVyfX3
Content-Disposition: form-data; name="myArray"
Content-Type: text/plain;charset=UTF-8
Content-Length: 3

One
--IpJ9ACFxVbhXlAQuIFccoJ0K0ttFS5PVyfX3
Content-Disposition: form-data; name="myArray"
Content-Type: application/json
Content-Length: 1

2
--IpJ9ACFxVbhXlAQuIFccoJ0K0ttFS5PVyfX3
Content-Disposition: form-data; name="myArray"
Content-Type: application/json
Content-Length: 4

true
--IpJ9ACFxVbhXlAQuIFccoJ0K0ttFS5PVyfX3
Content-Disposition: form-data; name="myArray"
Content-Type: text/plain;charset=UTF-8
Content-Length: 4

Four
--IpJ9ACFxVbhXlAQuIFccoJ0K0ttFS5PVyfX3--

```

Fixes #10876

## Type of change

- New feature (non-breaking change which adds functionality)
- This change requires a documentation update

## How Has This Been Tested?

- Unit test - ‍‍‍‍`DataUtilsTest.testParseMultipartArrayDataWorks()`

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

## Test coverage results :test_tube:
<details><summary>:white_circle: Total coverage has not changed</summary>

    // Code coverage diff between base branch:release and head branch: feature/10876-backend-allow_dynamic_string_arrays_multipart 
Status | File | % Stmts | % Branch | % Funcs | % Lines 
 -----|-----|---------|----------|---------|------ 
 :green_circle: | app/client/src/utils/autocomplete/TernServer.ts | 52.94 **(0.23)** | 41.67 **(0.84)** | 36.21 **(0)** | 56.99 **(0.25)**</details>

### CODE REVIEW

**Review by AmanAgarwal041 — DISMISSED** [review]

File: app/server/appsmith-plugins/restApiPlugin/src/main/java/com/external/helpers/DataUtils.java [hunk 1]
```diff
@@ -156,25 +159,49 @@ public String parseFormData(List<Property> bodyFormData, Boolean encodeParamsTog
...
+                        switch (multipartFormDataType) {
+                            case TEXT:
+                                byte[] valueBytesArray = new byte[0];
+                                if (StringUtils.hasLength(String.valueOf(property.getValue()))) {
+                                    valueBytesArray = String.valueOf(property.getValue()).getBytes(StandardCharsets.ISO_8859_1);
+                                }
+                                bodyBuilder.part(
+                                        key,
+                                        valueBytesArray,
+                                        MediaType.TEXT_PLAIN);                                break;
```
Comment (nidhi-nair) [inline app/server/appsmith-plugins/restApiPlugin/src/main/java/com/external/helpers/DataUtils.java:172]: Nit: Please format the file before commiting

File: app/server/appsmith-plugins/restApiPlugin/src/main/java/com/external/helpers/DataUtils.java [hunk 2]
```diff
@@ -156,25 +159,49 @@ public String parseFormData(List<Property> bodyFormData, Boolean encodeParamsTog
...
+                                } catch (IOException e) {
+                                    e.printStackTrace();
+                                    throw new AppsmithPluginException(
+                                            AppsmithPluginError.PLUGIN_DATASOURCE_ARGUMENT_ERROR,
+                                            "Unable to parse content. Expected to receive an array or object of multipart data"
+                                    );
+                                }
+                                break;
+                            case ARRAY:
+                                System.out.println(property);
```
Comment (nidhi-nair) [inline app/server/appsmith-plugins/restApiPlugin/src/main/java/com/external/helpers/DataUtils.java:185]: Can we remove this extraneous log?

File: app/server/appsmith-plugins/restApiPlugin/src/test/java/com/external/helpers/DataUtilsTest.java [hunk 3]
```diff
@@ -257,4 +259,43 @@ public void testParseMultipartFileData_withNullKeys_skipsNullProperty() {
...
+        Mono<Void> result = bodyInserter.insert(request, this.context);
+        StepVerifier.create(result).expectComplete().verify();
+        StepVerifier.create(DataBufferUtils.join(request.getBody()))
+                .consumeNextWith(dataBuffer -> {
+                    byte[] resultBytes = new byte[dataBuffer.readableByteCount()];
+                    dataBuffer.read(resultBytes);
+                    DataBufferUtils.release(dataBuffer);
+                    String content = new String(resultBytes, StandardCharsets.UTF_8);
+
+                    Assertions.assertThat(content).contains("Content-Disposition: form-data; name=\"arrayOne\"");
```
Comment (nidhi-nair) [inline app/server/appsmith-plugins/restApiPlugin/src/test/java/com/external/helpers/DataUtilsTest.java:292]: Can we test for each of the array values getting passed here?

**Review by nidhi-nair — CHANGES_REQUESTED** [review]

File: app/server/appsmith-plugins/restApiPlugin/src/main/java/com/external/helpers/DataUtils.java (hunk 1 above)
Comment (mojtab23) [reply app/server/appsmith-plugins/restApiPlugin/src/main/java/com/external/helpers/DataUtils.java]: Fixed!

File: app/server/appsmith-plugins/restApiPlugin/src/main/java/com/external/helpers/DataUtils.java (hunk 2 above)
Comment (mojtab23) [reply app/server/appsmith-plugins/restApiPlugin/src/main/java/com/external/helpers/DataUtils.java]: Fixed!

File: app/server/appsmith-plugins/restApiPlugin/src/test/java/com/external/helpers/DataUtilsTest.java (hunk 3 above)
Comment (mojtab23) [reply app/server/appsmith-plugins/restApiPlugin/src/test/java/com/external/helpers/DataUtilsTest.java]: I changed the test.

**Review by nidhi-nair — APPROVED** [review]
Looks good to me. @mojtab23, please also raise a Github issue for adding documentation for this change and assign it to @Pranay105. You will also need to provide a draft that they can then edit to add in the docs.

**Review by Irongade — APPROVED** [review]

**Pull request merged**